server/src/main/java/org/opensearch/index/engine/NRTReplicationEngine.java

```diff
@@ -0,0 +1,465 @@
+import java.util.function.BiFunction;
+import java.util.function.LongConsumer;
+import java.util.function.LongSupplier;
+import java.util.stream.Stream;
+
+/**
+ * This is an {@link Engine} implementation intended for replica shards when Segment Replication
+ * is enabled.  This Engine does not create an IndexWriter, rather it refreshes a {@link NRTReplicationReaderManager}
+ * with new Segments when received from an external source.
+ *
+ * @opensearch.internal
+ */
+public class NRTReplicationEngine extends Engine {
+
+    private volatile SegmentInfos lastCommittedSegmentInfos;
+    private final NRTReplicationReaderManager readerManager;
+    private final CompletionStatsCache completionStatsCache;
+    private final LocalCheckpointTracker localCheckpointTracker;
+    private final Translog translog;
+
+    public NRTReplicationEngine(EngineConfig engineConfig) {
+        super(engineConfig);
+        store.incRef();
+        NRTReplicationReaderManager readerManager = null;
+        try {
+            lastCommittedSegmentInfos = store.readLastCommittedSegmentsInfo();
+            readerManager = new NRTReplicationReaderManager(OpenSearchDirectoryReader.wrap(getDirectoryReader(), shardId));
+            final SequenceNumbers.CommitInfo commitInfo = SequenceNumbers.loadSeqNoInfoFromLuceneCommit(
+                this.lastCommittedSegmentInfos.getUserData().entrySet()
+            );
+            this.localCheckpointTracker = new LocalCheckpointTracker(commitInfo.maxSeqNo, commitInfo.localCheckpoint);
+            this.completionStatsCache = new CompletionStatsCache(() -> acquireSearcher("completion_stats"));
+            this.readerManager = readerManager;
+            this.readerManager.addListener(completionStatsCache);
+            this.translog = openTranslog(
+                engineConfig,
+                getTranslogDeletionPolicy(engineConfig),
+                engineConfig.getGlobalCheckpointSupplier(),
+                localCheckpointTracker::markSeqNoAsPersisted
+            );
+        } catch (IOException e) {
+            IOUtils.closeWhileHandlingException(store::decRef, readerManager);
+            throw new EngineCreationFailureException(shardId, "failed to create engine", e);
+        }
+    }
+
+    public synchronized void updateSegments(final SegmentInfos infos, long seqNo) throws IOException {
+        // Update the current infos reference on the Engine's reader.
+        readerManager.updateSegments(infos);
+
+        // only update the persistedSeqNo and "lastCommitted" infos reference if the incoming segments have a higher
+        // generation. We can still refresh with incoming SegmentInfos that are not part of a commit point.
+        if (infos.getGeneration() > lastCommittedSegmentInfos.getGeneration()) {
+            this.lastCommittedSegmentInfos = infos;
+        }
+        localCheckpointTracker.fastForwardProcessedSeqNo(seqNo);
+        readerManager.maybeRefresh();
```

Comment:
Do we need to call `maybeRefresh()` here because `readerManager.updateSegments(infos)` already does it https://github.com/opensearch-project/OpenSearch/pull/3240/files#diff-b27186aa9cf59bb1f1a2de55cb1871123611ea6366a0e28a11c163277f2cadfbR78

Reply:
good catch, this is redundant. Removed.